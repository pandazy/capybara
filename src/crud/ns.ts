import { ns as parentNs } from '../ns';

export function ns(context: string) {
  return parentNs(`crud:${context}`);
}
