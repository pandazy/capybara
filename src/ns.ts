export function ns(context: string) {
  return `Capybara:${context}`;
}
