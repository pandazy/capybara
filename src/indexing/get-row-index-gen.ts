import { CommonMap, Row } from '../types';
import { makeAssertUniquePass } from './make-assert-unique-pass';
import { hashByDefault, getIndexKeyGen } from './get-index-key-gen';
import { ns } from './ns';
import { getUniqCheck, UniqCheck } from './get-uniq-check';
import { Index, IndexDef } from './types';
import { keys } from '@pandazy/mole-core/dist/object-methods';

const Context = ns('getRowIndexGen');

export function getRowIndexGen<TRow extends Row>(
	indexDefMap: CommonMap<IndexDef>,
	hash = hashByDefault,
) {
	const indexers = keys(indexDefMap).map((defKey) => {
		const indexDef = indexDefMap[defKey];
		return {
			makeKey: getIndexKeyGen<TRow>(indexDef.fields, hash),
			uniqCheck: getUniqCheck<TRow>(indexDef, hash),
			indexDef,
		};
	});
	const assertUnique = makeAssertUniquePass(Context);
	return (index: Index, row: TRow, rowKey: string) => {
		const newIndexFragment: Index = indexers.reduce(
			(indexFragment, { makeKey, uniqCheck, indexDef }) => {
				assertUnique({
					row,
					indexDef,
					index,
					pass: uniqCheck as UniqCheck<Row>,
				});
				const indexKey = makeKey(row);
				const existingRowKeys = index[indexKey] || [];
				return {
					...indexFragment,
					[indexKey]: [...existingRowKeys, rowKey],
				};
			},
			{},
		);
		return {
			...index,
			...newIndexFragment,
		};
	};
}
