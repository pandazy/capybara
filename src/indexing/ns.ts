import { ns as parentNs } from '../ns';

export function ns(context: string) {
  return parentNs(`indexing:${context}`);
}
