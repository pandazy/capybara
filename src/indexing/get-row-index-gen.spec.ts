import { getRowIndexGen } from './get-row-index-gen';
import { ns } from './ns';

interface TestRow {
  name: string;
  job: string;
  species: string;
}

describe(ns('getRowIndexGen'), () => {
  it('should update index for a given row', () => {
    expect.assertions(2);

    const row: TestRow = {
      name: 'shinobu',
      job: 'Jobless',
      species: 'Vampire',
    };
    const fields1 = { fields: ['name', 'job'], isUnique: true };
    const fields2 = { fields: ['species', 'job'], isUnique: false };
    const indexDefs = {
      'job|name': fields1,
      'job|species': fields2,
    };
    const customHash = (val: string) => `%${val}%`;
    const index = {
      ['[job:knowall][name:tsubasa]']: ['3'],
      ['[job:jobless][species:vampire]']: ['2'],
    };
    const indexWithHash = {
      '%[job:knowall][name:tsubasa]%': ['3'],
      '%[job:Jobless][species:Vampire]%': ['2'],
    };

    const addIndex = getRowIndexGen<TestRow>(indexDefs);
    const addIndexWithHash = getRowIndexGen<TestRow>(indexDefs, customHash);

    expect(addIndex(index, row, '1')).toEqual({
      ['[job:knowall][name:tsubasa]']: ['3'],
      ['[job:jobless][name:shinobu]']: ['1'],
      ['[job:jobless][species:vampire]']: ['2', '1'],
    });
    expect(addIndexWithHash(indexWithHash, row, '1')).toEqual({
      '%[job:knowall][name:tsubasa]%': ['3'],
      '%[job:Jobless][name:shinobu]%': ['1'],
      '%[job:Jobless][species:Vampire]%': ['2', '1'],
    });
  });

  it('should throw an error when insert duplicate row into unique-indexed row', () => {
    expect.assertions(2);

    const row: TestRow = {
      name: 'shinobu',
      job: 'Jobless',
      species: 'Vampire',
    };

    const customHash = (val: string) => `%${val}%`;
    const index = {
      ['[job:jobless][name:shinobu]']: ['1'],
    };
    const indexWithHash = {
      '%[job:Jobless][name:shinobu]%': ['1'],
    };

    const fields1 = { fields: ['name', 'job'], isUnique: true };
    const indexDefs = {
      'job|name': fields1,
    };
    const addIndex = getRowIndexGen<TestRow>(indexDefs);
    const addIndexWithHash = getRowIndexGen<TestRow>(indexDefs, customHash);

    expect(() => addIndex(index, row, '1')).toThrowErrorMatchingSnapshot();
    expect(() =>
      addIndexWithHash(indexWithHash, row, '1'),
    ).toThrowErrorMatchingSnapshot();
  });
});
