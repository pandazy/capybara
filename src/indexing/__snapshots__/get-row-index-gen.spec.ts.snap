// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capybara:indexing:getRowIndexGen should throw an error when insert duplicate row into unique-indexed row 1`] = `
"
[context]: Capybara:indexing:getRowIndexGen
[failure cause]: duplicate row is not allowed
[actual value]: {\\"row\\":{\\"name\\":\\"shinobu\\",\\"job\\":\\"Jobless\\",\\"species\\":\\"Vampire\\"},\\"indexDef\\":{\\"fields\\":[\\"job\\",\\"name\\"],\\"isUnique\\":true},\\"index\\":{\\"[job:jobless][name:shinobu]\\":[\\"1\\"]}}
[actual value type]: Object"
`;

exports[`Capybara:indexing:getRowIndexGen should throw an error when insert duplicate row into unique-indexed row 2`] = `
"
[context]: Capybara:indexing:getRowIndexGen
[failure cause]: duplicate row is not allowed
[actual value]: {\\"row\\":{\\"name\\":\\"shinobu\\",\\"job\\":\\"Jobless\\",\\"species\\":\\"Vampire\\"},\\"indexDef\\":{\\"fields\\":[\\"job\\",\\"name\\"],\\"isUnique\\":true},\\"index\\":{\\"%[job:Jobless][name:shinobu]%\\":[\\"1\\"]}}
[actual value type]: Object"
`;
